lib/fir/fir-builder.h

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_IR_BUILDER_H_
+#define FORTRAN_IR_BUILDER_H_
+
+#include "fir-basicblock.h"
+#include "statements.h"
+#include <initializer_list>
+
+namespace Fortran::IR {
+
+struct FIRBuilder {
+  explicit FIRBuilder(BasicBlock &block)
+    : cursorRegion{block.getParent()}, cursorBlock{&block} {}
+  template<typename A> Statement &Insert(A *s);
+  template<typename A> Statement &InsertTerminator(A *s);
+  void SetInsertionPoint(BasicBlock *bb);
```

Comment:
Does this set the "insertion point" to the end of the block?  Would it precede any terminating statement?  Can one set the insertion point before/after any other statement?

Reply:
Yes. Conventionally, the constructing visitation will build the IR in a lexical order. However, yes, the LLVM builder also supports repositioning the cursor.